Flaky test: TestDockerDaemonSuite/TestDaemonRestartUnlessStopped (again)

this test was flaky before https://github.com/moby/moby/issues/38720, which was fixed in https://github.com/moby/moby/pull/38737

However, recently, it became flaky again. 

It was noticed as "flaky" on https://github.com/moby/moby/pull/41317#issuecomment-669834399 (not sure if it was noticed as flaky before that, so mentioning here, just in case it's a regression).

https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-41335/1/tests

```
=== RUN   TestDockerDaemonSuite/TestDaemonRestartUnlessStopped
    --- FAIL: TestDockerDaemonSuite/TestDaemonRestartUnlessStopped (6.33s)
        docker_cli_daemon_test.go:140: assertion failed: true (bool) != false (shouldRun bool): After daemon restart: container "exit" is running
```

--- a/daemon/kill.go
+++ b/daemon/kill.go
@@ -99,7 +99,19 @@ func (daemon *Daemon) killWithSignal(container *containerpkg.Container, sig int)
 		if errdefs.IsNotFound(err) {
 			unpause = false
 			logrus.WithError(err).WithField("container", container.ID).WithField("action", "kill").Debug("container kill failed because of 'container not found' or 'no such process'")
-			go daemon.handleContainerExit(container, nil)
+			go func() {
+				// We need to clean up this container but it is possible there is a case where we hit here before the exit event is processed
+				// but after it was fired off.
+				// So let's wait the container's stop timeout amount of time to see if the event is eventually processed.
+				// Doing this has the side effect that if no event was ever going to come we are waiting a a longer period of time uneccessarily.
+				// But this prevents race conditions in processing the container.
+				ctx, cancel := context.WithTimeout(context.TODO(), time.Duration(container.StopTimeout())*time.Second)
+				defer cancel()
+				s := <-container.Wait(ctx, containerpkg.WaitConditionNotRunning)
+				if s.Err() != nil {
+					daemon.handleContainerExit(container, nil)
+				}
+			}()
 		} else {
 			return errors.Wrapf(err, "Cannot kill container %s", container.ID)
 		}

